Test helper refactoring 2: Refactor and move functionality for generating call strings used in error messages

namurphy opened the pull request:
The second spin-off PR from #726.  This should be reviewed and merged after #919.

We currently have a function `call_string` which generates a string that can reproduce a particular call.  This function is currently used in test helper functionality when generating error messages.  Its potential uses go beyond creating error messages, though, such as in `__repr__` methods.  This PR moves that functionality to `plasmapy.utils.formatting`, and creates functions for reproducing the call strings of functions and attributes/methods of class instances.

namurphy commented:
- I removed colorama as a dependency, in part because it sometimes led to weird formatting in terminals.  
 - I sorted imports with `isort` and reformatted with `black` for a small number of files elsewhere in the package.

rocco8773 commented on plasmapy/utils/exceptions.py:
```diff
@@ -93,6 +94,14 @@ class PlasmaPyWarning(Warning):
     pass
 
 
+class PlasmaPyDeprecationWarning(DeprecationWarning, PlasmaPyWarning):
+    """
+    A warning for features in PlasmaPy that have been deprecated.
+    """
+
+    pass
+
+
```
What's the motivation for this, versus just using `DeprecationWarning`?  If we really want our own `DeprecationWarning`, then I would encourage considering an approach similar to how `sphinx` does it.   They do this [this](https://github.com/sphinx-doc/sphinx/blob/64498ad0dbab86fa42f5556d7bc84baf719cefda/sphinx/deprecation.py#L21-L33)...

```python
class RemovedInSphinx40Warning(DeprecationWarning):
    pass


class RemovedInSphinx50Warning(PendingDeprecationWarning):
    pass


class RemovedInSphinx60Warning(PendingDeprecationWarning):
    pass
```

So, our version would look like...

```python
class RemovedInPlasmaPy05Warning(DeprecationWarning, PlasmaPyWarning):
    pass


class RemovedInPlasmaPy06Warning(PendingDeprecationWarning, PlasmaPyWarning):
    pass
```

Now, I also think we should consider `FutureWarning` versus `DeprecationWarning`.  According to [Python's documentation](https://docs.python.org/3/library/warnings.html?highlight=warning#warning-categories), `DeprecationWarning` and `FutureWarnings` both announce deprecated features, but `DeprecationWarning` is intended for developers and `FutureWarning` is intended for end users.  Since we are constantly changing and deprecating end user features, we might want to consider using `FutureWarning` instead.

namurphy replied:
My original motivation was to follow Astropy's strategy for handling deprecations.  The benefit of having `PlasmaPyDeprecationWarning` is to have a bit more specificity than `DeprecationWarning`.  I'm more inclined to specify the version in which something is deprecated in a `@deprecated` decorator (#945) rather than having a series of different classes.  I don't have a strong opinion though.

I think I'll back off on the deprecation stuff in this PR since it's off-topic, but I'll link to this discussion in #945 since it's really relevant there.

rocco8773 commented on docs/utils/index.rst:
```diff
@@ -17,16 +17,17 @@ means:
  * the many kinds of warnings and exceptions you may (hopefully not!)
    encounter while working with PlasmaPy, such as
-   `plasmapy.utils.RelativityWarning` or `plasmapy.utils.PhysicsError`.
+   `~plasmapy.utils.RelativityWarning` or `~plasmapy.utils.PhysicsError`.
```
```suggestion
   `~plasmapy.utils.exceptions.RelativityWarning` or `~plasmapy.utils.exceptions.PhysicsError`.
```

Since we do not `automodapi` or `automodule` `plasmapy.utils`,  nothing defined in the `__all__` of `plasmapy/utils/__init__.py` is documented; thus, there is no stub file for things like `plasmapy.utils.RelativityWarning`.

rocco8773 commented on docs/utils/index.rst:
```diff
@@ -17,16 +17,17 @@ means:
-   checking, such as `plasmapy.utils.decorators.validate_quantities` and
-   `plasmapy.utils.check_relativistic`.
+   checking, such as `~plasmapy.utils.decorators.validate_quantities` and
+   `~plasmapy.utils.check_relativistic`.
```
```suggestion
   `~plasmapy.utils.decorators.check_relativistic`.
```

rocco8773 reviewed (changes requested):
This is just a partial review.  I'm only focusing on the documentation aspect of the functionality.

rocco8773 commented on plasmapy/utils/formatting/__init__.py:
```diff
@@ -0,0 +1,7 @@
+"""Tools for formatting strings, including for error messages."""
```
I would argue that this sub-directory could/should be decomposed into a single `formatting.py` file, much like we did with `roman.py`.  I'd back off this if you see a significant expansion of the functionality contained within this sub-directory, but I do not see that at the moment.

I also don't think the name `formatting` quite represents what this functionality is doing.  Something like `repr` would better reflect what the functionality is doing.  Obviously, `repr` would be a poor name due to naming conflicts. ...`code_repr`, `obj2str`, `obj2repr`, `obj2code`, `obj2codestr` ...idk, none of these seem to be elegant or representative.

namurphy replied:
Good idea...I'll move this into a single file.  I can't think of anything else what would go into `formatting` at the moment, though I expect some things here and there will come up.  

I'll think about the name...thus far some variation of `code_repr` seems best.

rocco8773 reviewed (changes requested):
Another partial review since GitHub doesn't let me respond to previous comments while I have a pending review.  Come on GitHub, it's almost 2021!!

namurphy commented on plasmapy/utils/formatting/formatting.py:
```diff
@@ -0,0 +1,255 @@
+from astropy import units as u
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+__all__ = ["call_string", "attribute_call_string", "method_call_string"]
```
If there ever was a code review comment that captures the essence of @rocco8773 reviewing @namurphy's pull request, this is it. :1st_place_medal: :100: :+1:  :joy_cat: 

I better go check the other PRs in this series!

namurphy commented:
My to do list...
 - [ ] Move `formatting.py` to its own file in utils, rather than in a directory in utils
 - [ ] Decide on a name for `formatting.py`
 - [ ] Remove deprecation stuff from this PR as its out-of-scope
 - [ ] Put `__all__` above imports everywhere

namurphy commented:
@rocco8773 — thanks again for the review!  I applied most of the suggestions so it's ready for another round when you have a chance.  It'd be particularly helpful if you could double check that the `docs/api_static` stuff is being handled correctly.

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+    return "\n\n".join(warnings_with_messages)
+
+
+def call_string(f: Callable, args: Any = tuple(), kwargs: Dict = {}) -> str:
```
I'm trying to understand the purpose of `call_string`.  Is it used for actual `assert` tests or is used only for reporting info associated with test failures?  For example,

```python
pytest.fail(f"Call {cal_string(...)} was suppose to do blah, but did doogah.")
```

I don't know, this seems like a lot of complexity and we're traveling rather far from [KISS](https://en.wikipedia.org/wiki/KISS_principle).  This functionality works beautifully for simple function calls, but would get out of hand if you were trying to use it on a function with a signature like [`scipy.signal.find_peaks`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks.html) (random example that would plausible in a `plasmapy.analysis` function).

namurphy replied:
The main purpose of `call_string`, etc. is for reporting info associated with test failures.  Reading the actual function call that led to a test failure can be extremely helpful for figuring out what the problem is, and it can help speed up debugging too, especially for newcomers to the project.  These functions could also be used in `__repr__`, and I think there would likely be times when these could be used directly in tests too.  

The point you made about managing complexity is important.  I'll need to think about that more...

rocco8773 replied:
I'm curious if you could sidestep the complexity if you incorporate some functionality form `inspect` to do the heavy lifting, like [`getcallargs`](https://docs.python.org/3/library/inspect.html#inspect.getcallargs) in combination with [`formatargspec`](https://docs.python.org/3/library/inspect.html#inspect.formatargspec).

namurphy replied:
That looks promising!  I'll look into it!

namurphy replied:
I'm not sure if what's in `inspect` will be able to take care of all the use cases that are needed.  Since `call_string` is being used to report testing information, it needs to be able to handle situations where the call is invalid, since invalid calls will sometimes need to be reconstructed and reported back to the user.  That is apparently out of scope for `inspect`.  Hm...

rocco8773 reviewed (commented):
Only a review on `plamsapy.utils.code_repr`.


This is a general statement to consider for all functionality in `code_repr.py`.  Since the `formatting` lingo is being replaced with the `code_repr` lingo, any naming or reference to `format` should probably also be replaced with `code_repr`.

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+    if not isinstance(arg, u.Quantity):
+        raise TypeError("Expecting a Quantity.")
+
+    formatted = f"{repr(arg.value)}"
+
+    if isinstance(arg.value, np.ndarray):
+        formatted = "np." + formatted
```
Yes, agreed.

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+
+def _format_arg(arg) -> str:
+    """Transform an argument into a format as would appear in a function call."""
+    if hasattr(arg, "__name__"):
+        return arg.__name__
```
From the [built-in types page](https://docs.python.org/3/library/stdtypes.html#definition.__name__):

> `definition.__name__`
> 
>     The name of the class, function, method, descriptor, or generator instance.

I wasn't sure about this, so it was helpful to look it up.

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+        for base, power in zip(arg.unit.bases, arg.unit.powers):
+            if power == -1:
+                formatted += f"/u.{base}"
+            elif power == 1:
+                formatted += f"*u.{base}"
+            else:
+                formatted += f"*u.{base}**{power}"
```
This seems like a good case to create a test for, too.

namurphy replied:
Thinking more, this shouldn't come up since `u.some_unit ** 0` turns into `u.dimensionless_unscaled`.

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+    attribute and ``showmodule`` is `True`, then prepend the module
+    name if not in `builtins`.
+    """
+    obj_name = obj.__name__ if hasattr(obj, "__name__") else repr(obj)
+
+    if hasattr(obj, "__name__") and showmodule is True:
+        module_name = inspect.getmodule(obj).__name__  # ADD TEST
+        if module_name != "builtins":
+            obj_name = f"{module_name}.{obj_name}"  # ADD TEST
```
Looking back at this, maybe I could also simplify this using `__qualname__`.  Will refactor...

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+    """
+    Transform a `~astropy.units.Quantity` into a format as would appear
+    in a function call.
+    """
```
I decided to tentatively go with names like `_code_repr_of_lihjdfkh`.  Thanks!

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,257 @@
+        args_and_kwargs += f"{_format_arg(arg)}, "
+
+    for kwarg in kwargs:
+        args_and_kwargs += f"{_format_kw(kwarg)}={_format_arg(kwargs[kwarg])}, "
```
Tentatively renaming it to `_format_kw_name`.

namurphy commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,323 @@
+        A `tuple` or `list` containing positional arguments, or any
+        other type of `object` if there is only one positional argument,
+        to be used during instantiation of ``cls``
+
+
```
```suggestion

```

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,321 @@
+        closing = f", ...{closing_brackets})"
+        return to_comma_before_ellipsis + closing
+
+    if not isinstance(array, np.ndarray):
+        raise TypeError("Expecting an ndarray.")
```
This scenario is not covered in the tests.

namurphy replied:
Since these are private functions, it might be better to do EAFP rather than LBYL (following a code review comment I received on Astropy).  I think I'll delete these lines.

rocco8773 commented on plasmapy/utils/tests/test_code_repr.py:
```diff
@@ -0,0 +1,370 @@
+@pytest.mark.parametrize(
+    "quantity, expected",
+    [
+        (5.3 * u.m, "5.3*u.m"),
+        (5.4 / u.m, "5.4/u.m"),
+        (5.5 * u.m ** -2, "5.5*u.m**-2"),
+        (5.0 * u.dimensionless_unscaled, "5.0*u.dimensionless_unscaled"),
+        (np.array([3.5, 4.2]) * u.m ** -2.5, "np.array([3.5, 4.2])*u.m**-2.5"),
```
For completeness, maybe throw in a case that looks like `5.0 / u.m ** 3`.

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,321 @@
+    if hasattr(arg, "__name__"):
+        return arg.__name__
+
+    function_for_type = {
+        u.Quantity: _code_repr_of_quantity,
+        np.ndarray: _code_repr_of_ndarray,
+    }
```
Should probably add `_code_repr_...`'s for long lists, tuples, and dicts.  I'm guessing these cases will turn up quickly.

rocco8773 commented on plasmapy/utils/tests/test_code_repr.py:
```diff
@@ -0,0 +1,370 @@
+@pytest.mark.parametrize(
+    "arg, expected",
+    [
+        (1, "1"),
+        ("asdf", "'asdf'"),
+        (3.42, "3.42"),
+        ([3.42, 3.84], "[3.42, 3.84]"),
+        (4.2 * u.m, "4.2*u.m"),
+        (np.array([1, 2, 3]), "np.array([1, 2, 3])"),
+        (np.array([[1, 2], [3, 4]]), "np.array([[1, 2], [3, 4]])"),
+        (np.array([1.0, 2.0, 3.0]) * u.m / u.s, "np.array([1., 2., 3.])*u.m/u.s"),
```
How about test cases for lists, tuples, and dicts?

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,321 @@
+    for arg in args_collection:
+        args_and_kwargs += f"{_code_repr_of_arg(arg)}, "
+
+    for kwarg in kwargs.keys():
+        args_and_kwargs += f"{kwarg}={_code_repr_of_arg(kwargs[kwarg])}, "
```
```suggestion
    for kw_name, kw_val in kwargs.items():
        args_and_kwargs += f"{kw_name}={_code_repr_of_arg(kw_val)}, "
```

Just think this is a little more readable...but your decision to take it or leave it.

rocco8773 commented on plasmapy/utils/tests/test_code_repr.py:
```diff
@@ -0,0 +1,370 @@
+    the positional and keyword arguments, as they would appear in a
+    function call.
+    """
+    args_and_kwargs = _code_repr_of_args_and_kwargs(args, kwargs)
+    if args_and_kwargs != expected:
+        pytest.fail(
+            f"_code_repr_of_args_and_kwargs with the following arguments:\n"
+            f"  args:   {args}\n"
+            f"  kwargs: {kwargs}\n"
+            f"is not returning the expected string:\n"
+            f"  expected: {repr(expected)}\n"
+            f"  actual:   {repr(args_and_kwargs)}"
+        )
```
Testing and reporting failures this way seems like it's sidestepping the [`assert` mechanics](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement).  Why not do something more like...

```suggestion
    args_and_kwargs = _code_repr_of_args_and_kwargs(args, kwargs)
    assert args_and_kwargs == expected, (
        f"_code_repr_of_args_and_kwargs with the following arguments:\n"
        f"  args:   {args}\n"
        f"  kwargs: {kwargs}\n"
        f"is not returning the expected string:\n"
        f"  expected: {repr(expected)}\n"
    )
```

rocco8773 replied:
I do think this approach should be applied to all similar cases.

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,321 @@
+
+def _name_with_article(ex: Exception) -> str:
+    """
+    Return a string with an indefinite article and name of
+    exception ``ex``.
```
Very few people here are going to be this familiar with grammar syntax, so this probably won't be clear for many.

```suggestion
    Return a string with an indefinite article and the name of
    the exception (e.g. `Exception` -> ``'an Exception'``.
```

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,321 @@
+    return f"{indefinite_article} {name}"
+
+
+def _substitute_module_shortcuts(module_name):
+    """Substitute common import shortcuts within module names."""
+    replacements = {
+        "numpy": "np",
+        "builtins": "",
+        "astropy.units.core": "u",
+        "astropy.units.quantity": "u",
+        "astropy.units.decorators": "u",
+        "astropy.units": "u",
+    }
+    for old, new in replacements.items():
+        if module_name.startswith(old):
+            module_name = module_name.replace(old, new, 1)
+    return module_name
```
If `_substitute_module_shortcuts()` is only going to be used by `_object_name()` then it's code should be contained with it `_object_name()`; otherwise, `_substitute_module_shortcuts()` should have its own dedicated tests.  If neither of those approaches are taken then these two functions are not 100% independent of each other can lead to problems down the line.

rocco8773 commented on plasmapy/utils/tests/test_code_repr.py:
```diff
@@ -0,0 +1,370 @@
+    (generic_function, ("x"), {}, "generic_function('x')"),
+    (generic_function, (1, "b", {}), {}, "generic_function(1, 'b', {})"),
+    (generic_function, (), {"kw": 1}, "generic_function(kw=1)"),
+    (generic_function, (), {"x": "c"}, "generic_function(x='c')"),
+    (
+        generic_function,
+        (1, 2, 3),
+        {"a": 4, "b": 5, "c": 6},
+        "generic_function(1, 2, 3, a=4, b=5, c=6)",
+    ),
+    (
+        generic_function,
+        np.array((1.0, 2.0)),
+        {},
+        "generic_function(np.array([1., 2.]))",
+    ),
+    (
+        generic_function,
+        np.array((3.0, 4.0)) * u.m,
+        {},
+        "generic_function(np.array([3., 4.])*u.m)",
+    ),
+    (
+        generic_function,
+        (ValueError,),
+        {"x": TypeError},
+        "generic_function(ValueError, x=TypeError)",
+    ),
+    (generic_function, (), {"bowl": super}, "generic_function(bowl=super)"),
+    (
+        generic_function,
+        (1, "b"),
+        {"b": 42, "R2": "D2"},
+        "generic_function(1, 'b', b=42, R2='D2')",
+    ),
+]
+
+
+@pytest.mark.parametrize("function,args,kwargs,expected", call_string_table)
+def test_call_string(function, args, kwargs, expected):
```
Unless `call_string_table` (and the like below) is (are) going to be called in several locations, I find that this disassociation of the parametrized list and the test function to significantly hinder readability of tests.  I mean, I start reading the test and then I have to wonder away from the test to get the cases that are being tested.

namurphy replied:
I think the main problem with readability here is that the when we parametrize tests using a list containing a series of tuples which might contain inner tuples, the parameters get detached from the corresponding semantic information/metadata.  It'd be so much more readable if the arguments were defined like keywords.  In #925 I define classes that are intended to address this problem...without me having to rewrite essentially the same code over and over again.  I might try doing something like that here, though independent of #925, since I'm trying to avoid using the test helpers to test the test helpers.

rocco8773 replied:
I agree with that sentiment.  `pytest.mark.parametrize`'s separation of names and values does make it a bit difficult to follow.

My point here is that when someone reads `@pytest.mark.parametrize("function,args,kwargs,expected", call_string_table)` they have to immediately search the code to find `call_string_table`.  The fact that I have to leave the test to find out what is being tested is what reduces readability for me.

rocco8773 commented on plasmapy/utils/code_repr.py:
```diff
@@ -0,0 +1,321 @@
+    cls_args : `tuple`, `list`, or any; optional
+        A `tuple` or `list` containing positional arguments, or any
+        other type of `object` if there is only one positional argument,
+        to be used during instantiation of ``cls``
+    cls_kwargs: `dict`, optional
```
```suggestion
        to be used during instantiation of ``cls``

    cls_kwargs: `dict`, optional
```

namurphy commented:
Changes:
 - I renamed `cls_args` to `args_to_cls` (and same to similar arguments)
 - I created namedtuples to make it easier to understand the things in parametrize
   - I thought...hey!  This is a job for dataclasses!  But it turns out, it wasn't.  Keep it simple splendorously.
 - I added some tests with a list or tuple as the argument.  I found a bug.  I fixed it.
 - I made some optional arguments keyword-only in order to enforce
 - I added a parameter `max_items` that can limit the number of items that will get printed out in an `ndarray` or a `Quantity`
 - Otherwise, I adopted most @rocco8773's suggestions.

rocco8773 reviewed (approved):
I'm satisfied. 😁 ... Looks good.  The `namedtuple` is an interesting approach to the problem, would not have thought about it myself.  I'm guessing you decided to leave the `code_repr` functions for lists, tuples, and dicts for a separate PR, correct?
